src/algorithms/ctmrg.jl

```diff
@@ -1,172 +1,381 @@
-    alg::CTMRG,
-    envs=CTMRGEnv(peps_above, peps_below),
-)
-    err = Inf
-    iter = 1
-
-    #for convergence criterium we use the on site contracted boundary
-    #this convergences, though the value depends on the bond dimension χ
-    old_norm = 1.0
-    new_norm = old_norm
-    ϵ₁ = 1.0
-    while (err > alg.tol && iter <= alg.maxiter) || iter <= alg.miniter
-        ϵ = 0.0
-        for i in 1:4
-            envs, ϵ₀ = left_move(peps_above, peps_below, alg, envs)
-            ϵ = max(ϵ, ϵ₀)
-            envs = rotate_north(envs, EAST)
-            peps_above = envs.peps_above
-            peps_below = envs.peps_below
+        conv_condition = max(Δnorm, ΔCS, ΔTS) < alg.tol && i > alg.miniter
+        ignore_derivatives() do # Print verbose info
+            if alg.verbosity > 1 || (alg.verbosity == 1 && (i == 1 || conv_condition))
+                @printf(
+                    "CTMRG iter: %3d   norm: %.2e   Δnorm: %.2e   ΔCS: %.2e   ΔTS: %.2e   ϵ: %.2e   Δϵ: %.2e\n",
+                    i,
+                    abs(normnew),
+                    Δnorm,
+                    ΔCS,
+                    ΔTS,
+                    ϵ,
+                    Δϵ
+                )
+            end
+            alg.verbosity > 0 &&
+                i == alg.maxiter &&
+                @warn(
+                    "CTMRG reached maximal number of iterations at (Δnorm=$Δnorm, ΔCS=$ΔCS, ΔTS=$ΔTS)"
+                )
+        end
+        conv_condition && break  # Converge if maximal Δ falls below tolerance
+
+        # Update convergence criteria
+        normold = normnew
+        CSold = CSnew
+        TSold = TSnew
+        ϵold = ϵ
+    end
+
+    # Do one final iteration that does not change the spaces
+    alg_fixed = CTMRG(;
+        alg.trscheme, alg.tol, alg.maxiter, alg.miniter, alg.verbosity, fixedspace=true
+    )
+    env′, = ctmrg_iter(state, env, alg_fixed)
+    envfix = gauge_fix(env, env′)
+    check_elementwise_convergence(env, envfix; atol=alg.tol^(3 / 4)) ||
+        @warn "CTMRG did not converge elementwise."
+    return envfix
+end
+
+"""
+    gauge_fix(envprev::CTMRGEnv{C,T}, envfinal::CTMRGEnv{C,T}) where {C,T}
+
+Fix the gauge of `envfinal` based on the previous environment `envprev`.
+This assumes that the `envfinal` is the result of one CTMRG iteration on `envprev`.
+Given that the CTMRG run is converged, the returned environment will be
+element-wise converged to `envprev`.
+"""
+function gauge_fix(envprev::CTMRGEnv{C,T}, envfinal::CTMRGEnv{C,T}) where {C,T}
+    # Check if spaces in envprev and envfinal are the same
+    same_spaces = map(Iterators.product(axes(envfinal.edges)...)) do (dir, r, c)
+        space(envfinal.edges[dir, r, c]) == space(envprev.edges[dir, r, c]) &&
+            space(envfinal.corners[dir, r, c]) == space(envprev.corners[dir, r, c])
+    end
+    @assert all(same_spaces) "Spaces of envprev and envfinal are not the same"
+
+    # Try the "general" algorithm from https://arxiv.org/abs/2311.11894
```

Comment:
I think in general that's a good idea but perhaps we should tackle that in a different PR. There are more sophisticated gauge fixing techniques (and also much simpler ones), so it might take some extra time to get that working. I'd suggest we leave it for now and come back later, when we need it ;-)